src/hotspot/share/gc/z/zRelocationSet.cpp

```diff
@@ -22,35 +22,102 @@
  */
 
 #include "precompiled.hpp"
-#include "gc/z/zForwarding.hpp"
+#include "gc/z/zArray.inline.hpp"
+#include "gc/z/zForwarding.inline.hpp"
+#include "gc/z/zForwardingAllocator.inline.hpp"
 #include "gc/z/zRelocationSet.hpp"
-#include "memory/allocation.hpp"
+#include "gc/z/zRelocationSetSelector.inline.hpp"
+#include "gc/z/zStat.hpp"
+#include "gc/z/zTask.hpp"
+#include "gc/z/zWorkers.hpp"
+#include "runtime/atomic.hpp"
 
-ZRelocationSet::ZRelocationSet() :
-    _forwardings(NULL),
-    _nforwardings(0) {}
+class ZRelocationSetInstallTask : public ZTask {
+private:
+  ZForwardingAllocator* const    _allocator;
+  ZForwarding**                  _forwardings;
+  const size_t                   _nforwardings;
+  ZArrayParallelIterator<ZPage*> _small_iter;
+  ZArrayParallelIterator<ZPage*> _medium_iter;
+  volatile size_t                _small_next;
+  volatile size_t                _medium_next;
+
+  void install(ZForwarding* forwarding, volatile size_t* next) {
+    const size_t index = Atomic::fetch_and_add(next, 1u);
+    _forwardings[index] = forwarding;
```

Comment:
Maybe `assert(index < _nforwardings)` before indexing the array?

Reply:
Will fix.